### ISSUE TITLE
network interface is renamed although NAME has been set by udev rule

### ISSUE THREAD

**Issue by mbiebl** [systemd/systemd#11436]
**systemd version the issue has been seen with**

v240

**Used distribution**

Debian sid
Downstream bug report at https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=919390

To reproduce the issue, create a file `/etc/udev/rules.d/70-persistent-net.rules` containing
```
SUBSYSTEM=="net", ACTION=="add", DRIVERS=="?*", ATTR{address}=="<MAC>", KERNEL=="eth*", NAME="lan0"
```
with the mac address of your ethernet network interface.
Unload the network module (in my case `8139cp`), then load it again.
Notice how the interface is properly renamed:
```
[ 3750.870434] 8139cp 0000:00:03.0 lan0: renamed from eth0
```
Now run `udevadm trigger --action=add`
```
[ 3752.509458] 8139cp 0000:00:03.0 ens3: renamed from lan0
```
The interface is renamed although a custom NAME has been set.

**Comment by mbiebl** [comment]
that's a regression compared to v239, and I'm inclined to add it to the v241 milestone.
@yuwata, @keszybz wdyt?

**Comment by mbiebl** [comment]
`udevadm info` after loading the kernel module:
```
P: /devices/pci0000:00/0000:00:03.0/net/lan0
L: 0
E: DEVPATH=/devices/pci0000:00/0000:00:03.0/net/lan0
E: INTERFACE=lan0
E: IFINDEX=3
E: SUBSYSTEM=net
E: USEC_INITIALIZED=4242831867
E: ID_NET_NAMING_SCHEME=v240
E: ID_NET_NAME_MAC=enx525400f34de9
E: ID_NET_NAME_PATH=enp0s3
E: ID_NET_NAME_SLOT=ens3
E: ID_BUS=pci
E: ID_VENDOR_ID=0x10ec
E: ID_MODEL_ID=0x8139
E: ID_PCI_CLASS_FROM_DATABASE=Network controller
E: ID_PCI_SUBCLASS_FROM_DATABASE=Ethernet controller
E: ID_VENDOR_FROM_DATABASE=Realtek Semiconductor Co., Ltd.
E: ID_MODEL_FROM_DATABASE=RTL-8100/8101L/8139 PCI Fast Ethernet Adapter (QEMU Virtual Machine)
E: ID_PATH=pci-0000:00:03.0
E: ID_PATH_TAG=pci-0000_00_03_0
E: ID_NET_DRIVER=8139cp
E: ID_NET_LINK_FILE=/usr/lib/systemd/network/99-default.link
E: ID_NET_NAME=ens3
E: SYSTEMD_ALIAS=/sys/subsystem/net/devices/lan0 /sys/subsystem/net/devices/lan0
E: TAGS=:systemd:
```
After running `udevadm trigger`:
```
P: /devices/pci0000:00/0000:00:03.0/net/ens3
L: 0
E: DEVPATH=/devices/pci0000:00/0000:00:03.0/net/ens3
E: INTERFACE=ens3
E: IFINDEX=3
E: SUBSYSTEM=net
E: USEC_INITIALIZED=4242831867
E: ID_NET_NAMING_SCHEME=v240
E: ID_NET_NAME_MAC=enx525400f34de9
E: ID_NET_NAME_PATH=enp0s3
E: ID_NET_NAME_SLOT=ens3
E: ID_BUS=pci
E: ID_VENDOR_ID=0x10ec
E: ID_MODEL_ID=0x8139
E: ID_PCI_CLASS_FROM_DATABASE=Network controller
E: ID_PCI_SUBCLASS_FROM_DATABASE=Ethernet controller
E: ID_VENDOR_FROM_DATABASE=Realtek Semiconductor Co., Ltd.
E: ID_MODEL_FROM_DATABASE=RTL-8100/8101L/8139 PCI Fast Ethernet Adapter (QEMU Virtual Machine)
E: ID_PATH=pci-0000:00:03.0
E: ID_PATH_TAG=pci-0000_00_03_0
E: ID_NET_DRIVER=8139cp
E: ID_NET_LINK_FILE=/usr/lib/systemd/network/99-default.link
E: ID_NET_NAME=ens3
E: SYSTEMD_ALIAS=/sys/subsystem/net/devices/ens3 /sys/subsystem/net/devices/ens3
E: TAGS=:systemd:
```

**Comment by mbiebl** [comment]
With v239
after modprobe:
```
P: /devices/pci0000:00/0000:00:03.0/net/lan0
E: DEVPATH=/devices/pci0000:00/0000:00:03.0/net/lan0
E: ID_BUS=pci
E: ID_MODEL_FROM_DATABASE=RTL-8100/8101L/8139 PCI Fast Ethernet Adapter (QEMU Virtual Machine)
E: ID_MODEL_ID=0x8139
E: ID_NET_DRIVER=8139cp
E: ID_NET_LINK_FILE=/usr/lib/systemd/network/99-default.link
E: ID_NET_NAME=ens3
E: ID_NET_NAME_MAC=enx525400f34de9
E: ID_NET_NAME_PATH=enp0s3
E: ID_NET_NAME_SLOT=ens3
E: ID_PATH=pci-0000:00:03.0
E: ID_PATH_TAG=pci-0000_00_03_0
E: ID_PCI_CLASS_FROM_DATABASE=Network controller
E: ID_PCI_SUBCLASS_FROM_DATABASE=Ethernet controller
E: ID_VENDOR_FROM_DATABASE=Realtek Semiconductor Co., Ltd.
E: ID_VENDOR_ID=0x10ec
E: IFINDEX=4
E: INTERFACE=lan0
E: SUBSYSTEM=net
E: SYSTEMD_ALIAS=/sys/subsystem/net/devices/lan0 /sys/subsystem/net/devices/lan0
E: TAGS=:systemd:
E: USEC_INITIALIZED=151929036
```
after `udevadm trigger --action=add`:
```
P: /devices/pci0000:00/0000:00:03.0/net/lan0
E: DEVPATH=/devices/pci0000:00/0000:00:03.0/net/lan0
E: ID_BUS=pci
E: ID_MODEL_FROM_DATABASE=RTL-8100/8101L/8139 PCI Fast Ethernet Adapter (QEMU Virtual Machine)
E: ID_MODEL_ID=0x8139
E: ID_NET_DRIVER=8139cp
E: ID_NET_LINK_FILE=/usr/lib/systemd/network/99-default.link
E: ID_NET_NAME_MAC=enx525400f34de9
E: ID_NET_NAME_PATH=enp0s3
E: ID_NET_NAME_SLOT=ens3
E: ID_PATH=pci-0000:00:03.0
E: ID_PATH_TAG=pci-0000_00_03_0
E: ID_PCI_CLASS_FROM_DATABASE=Network controller
E: ID_PCI_SUBCLASS_FROM_DATABASE=Ethernet controller
E: ID_VENDOR_FROM_DATABASE=Realtek Semiconductor Co., Ltd.
E: ID_VENDOR_ID=0x10ec
E: IFINDEX=4
E: INTERFACE=lan0
E: SUBSYSTEM=net
E: SYSTEMD_ALIAS=/sys/subsystem/net/devices/lan0
E: TAGS=:systemd:
E: USEC_INITIALIZED=151929036
```

**Comment by poettering** [comment]
Hmm, I wonder if this is intended effect of 55b6530baacf4658a183b15b010a8cf3483fde08

**Comment by poettering** [comment]
Also see #9006

**Comment by poettering** [comment]
And of course #9088

**Comment by poettering** [comment]
Not sure what the right approach here is

**Comment by yuwata** [comment]
Why it is named `lan0` when the driver is loaded?? If, the final result is `ens3` then I expect it should be always `ens3`.

**Comment by mbiebl** [comment]
It should always be named `lan0` because of the udev rule.

**Comment by yuwata** [comment]
What happens when the condition `KERNEL=="eth*"` is dropped?

**Comment by yuwata** [comment]
I've not tested that, but I guess that, because of the condition, 
- initially it is named to `lan0`, as the initial name is `ethX`.
- the later evaluation results `ens3`, as the interface is _not_ `ethX` anymore.

v239 or before, it is evaluated only once. So, the name was stayed at `lan0`. But, v240, 55b6530baacf4658a183b15b010a8cf3483fde08 makes the rule is always evaluated.

**Comment by yuwata** [comment]
If I remember correctly, the original motivation of the commit is that interface name can be changed by udev rules without updating initrd or changing kernel command line. So, if my above guess is correct,,, what should we do?

**Comment by mbiebl** [comment]
imho, revert and find another solution #9006

**Comment by mbiebl** [comment]
@yuwata a default policy like /lib/systemd/network/99-default.link should never trump explicit user configuration.

**Comment by keszybz** [comment]
@mbiebl There is no split between "explicit user configuration" and the rest. Files like `99-default` are named this way to have the lower priority. But there is no magic of "oh, this was set by the user, let's not touch it" anywhere.  The problem with trying to make a division like that is that nobody seems to ever agree what is "user" configuration, and what is "explicit".  Current approach of simply executing the configuration as we find it is much clearer and sustainable. #9006 was applied precisely because users were creating configuration and were peeved that some very specific parts of that configuration are not applied.

The rules in the bug report were relying on an implementation detail of the rule engine, and changing this implementation detail is something that we are allowed to do. The rule should never have had `kernel=="eth*"` if it was supposed to apply also after the interface was renamed. Current behaviour seems correct to me: when the device is called `lan0`, there's just one rule that applies to the devices, and when `udevadm trigger` fires off, we execute all rules that apply. This is also much nicer for the user, because they can create configuration and apply it, without jumping through hoops to reset the state.

**Comment by mbiebl** [comment]
It's not nicer to users, as it breaks existing user configurations. Which is ***bad***

**Comment by mbiebl** [comment]
I'm amazed that I have to point this out....

**Comment by keszybz** [comment]
Not every facet of program behaviour is guaranteed to be stable. Generally, only things that are documented are promised to be kept unchanged. Is the fact that link renames are applied only once documented anywhere?

You seem to suggest that we should never change any user visible detail. I think this user rule was in error, and it worked for a while by luck, and now it doesn't. This happens all the time.

**Comment by mbiebl** [comment]
sigh

**Comment by mbiebl** [comment]
I guess I'll stop filing bug reports

**Comment by vcaputo** [comment]
wouldn't it make sense to only apply `kernel=` matching to the kernel-assigned name, not a potentially renamed name?

**Comment by bl33pbl0p** [comment]
This doesn't seem to have ended well.

https://lists.freedesktop.org/archives/systemd-devel/2019-January/041971.html

**Comment by filbranden** [comment]
From the peanut gallery...

Would it make sense to somehow record the *original* name of this interface and match `KERNEL=="eth*"` against that name as well?

When reading the rule from the OP, that was my impression of how that should have worked (and the previous one-rename rule was a good approximation of it...)

Not really sure whether that makes sense, would be feasible and would address the other issues that commit was supposed to fix...

**Comment by poettering** [comment]
Urks, I find this breakage quite unfortunate I must say. I mean, writing a rule that matches effects of a previous rule (like the one debian is using there) is probably not a great idea (rules really should match stuff that is not going to change for a device, so that they are idempotent), but I am not sure that simply breaking systems like this is a good idea.

I don't have a strong opinion on the way out, but I am slightly leaning towards reverting the patch and simply asking downstreams to drop any rules that rename ifaces whatsoever from their initrds, as that means the host's policy will never be enforced. I mean, afaiu the patch was done precisely to ensure that host policy applies if initrds already renamed devices, no?

I mean, users/downstreams assumed that these rules where OK, and we probably should tell them they are a bad idea, but also, I think we shouldn't break so many installed systems this way...

Another idea might be make it configurable in the .link file whether to rename already renamed interfaces, with a default to on. Then, change our 99-default.link file to turn off the option. This way, .link files supplied by users will by default override everything, but the fallback one we provide won't.

Other ideas, opinions?

**Comment by fsateler** [comment]
> Other ideas, opinions?

The current documentation does not specify whether user modifications are read, and there are two factors suggesting they shouldn't:

1. The name
2. The existence of a NAME variable for matching against userspace changes.

Honestly, having a variable named KERNEL map to a value provided by userspace is weird.

**Comment by yuwata** [comment]
> Not every facet of program behaviour is guaranteed to be stable. Generally, only things that are documented are promised to be kept unchanged. Is the fact that link renames are applied only once documented anywhere?

Hmm, we have already documented that implicitly...
https://github.com/systemd/systemd/blob/850115b3a1119c87f14ecb7634615632ef0b1933/man/systemd.link.xml#L261-L262

At least, we need to update this.

**Comment by yuwata** [comment]
@poettering's proposal is implemented in #11443.

**Comment by keszybz** [comment]
> <literal>NAME</literal>. If the name has already been set by 
userspace, no renaming is performed. The available policies 

OK, that is enough for me to consider the previous behaviour documented. So I agree that we should preserve compatibility for this.

> Another idea might be make it configurable in the .link file whether to rename already renamed interfaces, with a default to on. Then, change our 99-default.link file to turn off the option. This way, .link files supplied by users will by default override everything, but the fallback one we provide won't.

I like this approach. The question is whether to default to "on" or "off" with current link files. If we default to "off", we preserve more backwards compatibility. If we default to "on", we get saner behaviour.

**Comment by yuwata** [comment]
Or, how about disabling the new behavior if naming scheme is specified earlier than v240?

**Comment by poettering** [comment]
@yuwata interesting idea. Maybe like this: if naming scheme is < v240 we default to RenameOnce=yes and otherwise to RenameOnce=no (and 99-default.link would set RenameOnce=yes either way)

**Comment by thecodingdude** [comment]
> Not every facet of program behaviour is guaranteed to be stable

Please go [talk](https://lkml.org/lkml/2012/12/23/75) to Linus Torvalds. If it works today and you make a change to break users, you are in the wrong. I should hope you wouldn't need yelling at to understand that.

Software developers need to internalize this and abide by it at all times, life would be much easier.

**Comment by bl33pbl0p** [comment]
Please don't post offtopic comments.

(This is making rounds elsewhere, locking this for now might be best).

**Comment by poettering** [comment]
BTW, for those coming late, a good fix (including a revert) has been merged now. I think all should be good now.
